service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/update/RetrieveControlledResourceMetadataStep.java

```diff
@@ -0,0 +1,62 @@
+package bio.terra.workspace.service.resource.controlled.flight.update;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.db.ResourceDao;
+import bio.terra.workspace.service.resource.WsmResource;
+import bio.terra.workspace.service.resource.controlled.ControlledResource;
+import bio.terra.workspace.service.resource.model.StewardshipType;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import java.util.UUID;
+
+/**
+ * Save a copy of the metadata before update so that the update can be restored on the undo path.
+ * Uses the working map of the flight
+ */
+public class RetrieveControlledResourceMetadataStep implements Step {
+
+  private final ResourceDao resourceDao;
+  private final UUID workspaceId;
+  private final UUID resourceId;
+
+  public RetrieveControlledResourceMetadataStep(
+      ResourceDao resourceDao, UUID workspaceId, UUID resourceId) {
+    this.resourceDao = resourceDao;
+    this.workspaceId = workspaceId;
+    this.resourceId = resourceId;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    WsmResource resource = resourceDao.getResource(workspaceId, resourceId);
+    if (resource.getStewardshipType() != StewardshipType.CONTROLLED) {
+      throw new IllegalArgumentException(
```

Comment:
Suggestion finding a better exception or make one. We shouldn't be throwing `IllegalArgument` or `IllegalState` in the mainline code as a general rule.

And this is yet another case where the throw from the cast call is the wrong exception. Sigh... I got that wrong big time...

Reply:
The throw will be caught by Stairway and will be the return error for the flight. Since `Illegal*` isn't derived from one of the common exception base classes, it gets whatever default the global exception handler gives it.

I'd prefer to exercise more control at the point of exception. See `InvalidMetadataException` as an example of a "totally server side" error.